cocotb/share/lib/embed/gpi_embed.c

```diff
@@ -243,124 +268,154 @@ int embed_sim_init(gpi_sim_info_t *info)
+        goto cleanup;
+    }
     for (i = 0; i < info->argc; i++) {
-        arg_value = PyString_FromString(info->argv[i]);
-        PyList_Append(argv_list, arg_value);
+        argv_item = PyString_FromString(info->argv[i]);                 // New reference
+        if (argv_item == NULL) {
+            PyErr_Print();
+            LOG_ERROR("Unable to create Python string from argv[%d] = \"%s\"", i, info->argv[i]);
+            Py_DecRef(argv_list);
+            goto cleanup;
+        }
+        if (-1 == PyList_SetItem(argv_list, i, argv_item))              // Note: This function steals the reference to argv_item
+        {
+            PyErr_Print();
+            LOG_ERROR("Unable to set argv list item");
+            Py_DecRef(argv_list);
+            goto cleanup;
+        }
     }
 
-    arg_dict = PyModule_GetDict(cocotb_module);
-    PyDict_SetItemString(arg_dict, "argv", argv_list);
-
-    argc = PyInt_FromLong(info->argc);
-    PyDict_SetItemString(arg_dict, "argc", argc);
+    // Add argv list to cocotb module
+    if (-1 == PyModule_AddObject(cocotb_module, "argv", argv_list)) {   // Note: This function steals the reference to argv_list if successful
+        PyErr_Print();
+        LOG_ERROR("Unable to set argv");
+        Py_DecRef(argv_list);
+        goto cleanup;
+    }
 
-    if (!PyCallable_Check(simlog_func)) {
+    // Add argc to cocotb module
+    if (-1 == PyModule_AddIntConstant(cocotb_module, "argc", info->argc)) {
         PyErr_Print();
-        LOG_ERROR("_printRecord is not callable");
+        LOG_ERROR("Unable to set argc");
         goto cleanup;
     }
 
     LOG_INFO("Running on %s version %s", info->product, info->version);
     LOG_INFO("Python interpreter initialized and cocotb loaded!");
 
     // Now that logging has been set up ok we initialize the testbench
-    if (-1 == PyObject_SetAttrString(cocotb_module, "SIM_NAME", PyString_FromString(info->product))) {
+    if (-1 == PyModule_AddStringConstant(cocotb_module, "SIM_NAME", info->product)) {
         PyErr_Print();
         LOG_ERROR("Unable to set SIM_NAME");
         goto cleanup;
     }
 
-    if (-1 == PyObject_SetAttrString(cocotb_module, "SIM_VERSION", PyString_FromString(info->version))) {
-            PyErr_Print();
-            LOG_ERROR("Unable to set SIM_VERSION");
-            goto cleanup;
+    if (-1 == PyModule_AddStringConstant(cocotb_module, "SIM_VERSION", info->version)) {
+        PyErr_Print();
+        LOG_ERROR("Unable to set SIM_VERSION");
+        goto cleanup;
     }
 
     // Set language in use as an attribute to cocotb module, or None if not provided
     const char *lang = getenv("TOPLEVEL_LANG");
     PyObject* PyLang;
     if (lang)
-        PyLang = PyString_FromString(lang);
-    else
-        PyLang = Py_None;
+        PyLang = PyString_FromString(lang);         // New reference
+    else {
+        PyLang = Py_BuildValue("");                 // New reference to None object
```

Comment:
This is a very strange spelling - I'd be inclined to use the faster
```C
Py_INCREF(Py_None);
PyLang = Py_None;
```

Reply:
Changed.